File: java/src/main/java/world/md2html/plugins/IncludeFilePlugin.java
package world.md2html.plugins;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.networknt.schema.JsonSchema;
import world.md2html.options.argfile.ArgFileParseException;
import world.md2html.options.model.Document;
import world.md2html.options.model.SessionOptions;
import world.md2html.pagemetadata.PageMetadataHandlersWrapper;
import world.md2html.utils.JsonUtils;
import world.md2html.utils.SmartSubstringer;
import world.md2html.utils.UserError;
import world.md2html.utils.Utils;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import static world.md2html.plugins.PluginUtils.mapFromStringOrObject;
import static world.md2html.utils.JsonUtils.loadJsonSchemaFromResource;
import static world.md2html.utils.Utils.getCachedString;
import static world.md2html.utils.Utils.supplyWithFileExceptionAsUserError;

public class IncludeFilePlugin extends AbstractMd2HtmlPlugin implements PageMetadataHandler {

    private static class IncludeFileData {
        private String rootDir = "";
        private String trim = "all";
        private boolean recursive = true;
        private SmartSubstringer substringer;
    }

    private Map<String, IncludeFileData> data;
    private PageMetadataHandlersWrapper metadataHandlers;
    private JsonSchema metadataSchema =
            loadJsonSchemaFromResource("plugins/include_file_metadata_schema.json");

    @Override
    public void acceptData(JsonNode data) throws ArgFileParseException {
        assureAcceptDataOnce();
        validateInputDataAgainstSchemaFromResource(data, "plugins/include_file_schema.json");

        Map<String, IncludeFileData> dataMap = new HashMap<>();
        for (JsonNode item : data) {
            ObjectNode itemNode = (ObjectNode) item;
            for (JsonNode jsonNode : itemNode.get("markers")) {
                String marker = jsonNode.asText().toUpperCase();
                if (dataMap.containsKey(marker)) {
                    throw new UserError("Marker duplication (case-insensitively): " + marker);
                }
                IncludeFileData includeFileData = new IncludeFileData();
                includeFileData.rootDir = itemNode.get("root-dir").asText();
                includeFileData.trim = itemNode.has("trim") ?
                        itemNode.get("trim").asText("all") : "all";
                includeFileData.recursive = itemNode.has("recursive") &&
                        itemNode.get("recursive").asBoolean();
                includeFileData.substringer = new SmartSubstringer(
                        itemNode.has("start-with") ? itemNode.get("start-with").asText("") : "",
                        itemNode.has("end-with") ? itemNode.get("end-with").asText("") : "",
                        itemNode.has("start-marker") ? itemNode.get("start-marker").asText("") : "",
                        itemNode.has("end-marker") ? itemNode.get("end-marker").asText("") : ""
                );
                dataMap.put(marker, includeFileData);
            }
            this.data = dataMap;
        }
    }

    @Override
    public boolean isBlank() {
        return data.isEmpty();
    }

    @Override
    public void acceptAppData(SessionOptions options, List<Md2HtmlPlugin> plugins,
                              PageMetadataHandlersWrapper metadataHandlers) {
        this.metadataHandlers = metadataHandlers;
    }

    @Override
    public List<PageMetadataHandlerInfo> pageMetadataHandlers() {
        return this.data.keySet().stream().map(marker ->
                new PageMetadataHandlerInfo(this, marker, false)).collect(Collectors.toList());
    }

    @Override
    public String acceptPageMetadata(Document document, String marker, String metadata,
                                     String metadataSection, Set<String> visitedMarkers
    ) throws PageMetadataException {

        IncludeFileData markerData = this.data.get(marker);
        SmartSubstringer substringer = markerData.substringer;

        Map<String, Object> metadataMap;
        try {
            //noinspection unchecked
            metadataMap = (Map<String, Object>)
                    JsonUtils.deJson(mapFromStringOrObject(metadata.trim(),
                    "file", this.metadataSchema));
        } catch (UserError e) {
            throw new UserError("Error in inclusion: " + e.getMessage() + ", page: '" +
                    document.getInput());
        }

        substringer = substringer.smartCopy(
                (String) metadataMap.get("start-with"),
                (String) metadataMap.get("end-with"),
                (String) metadataMap.get("start-marker"),
                (String) metadataMap.get("end-marker")
        );
        String filePath = ((String) metadataMap.get("file")).trim();
        Path includeFile = Paths.get(markerData.rootDir, filePath);
        String content = supplyWithFileExceptionAsUserError(
                () -> getCachedString(includeFile, Utils::readStringFromUtf8File),
                "Error processing page metadata block"
        );

        content = substringer.substring(content);

        String trim = (String) metadataMap.getOrDefault("trim", markerData.trim);
        if ("all".equals(trim)) {
            content = content.trim();
        } else if ("empty-lines".equals(trim)) {
            content = Utils.stripEmptyLines(content);
        }

        boolean recursive;
        if (metadataMap.containsKey("recursive")) {
            recursive = (Boolean) metadataMap.get("recursive");
        } else {
            recursive = markerData.recursive;
        }
        return recursive ?
                metadataHandlers.applyMetadataHandlers(content, document, visitedMarkers,
                        "INCLUDE_FILE_PLUGIN:" + includeFile) :
                content;
    }
}

File: python/plugins/include_file_plugin.py
import json
from pathlib import Path
from typing import Dict, Union

from models.document import Document
from models.options import Options
from models.page_metadata_handlers import PageMetadataHandlers
from page_metadata_utils import apply_metadata_handlers
from plugins.md2html_plugin import Md2HtmlPlugin
from plugins.plugin_utils import dict_from_string_or_object
from utils import read_lines_from_cached_file, UserError, SmartSubstringer, strip_empty_lines

MODULE_DIR = Path(__file__).resolve().parent


class IncludeFileData:
    def __init__(self, root_dir: str, trim: Union[str, None], recursive: bool,
                 # start_with: str, end_with: str, start_marker: str, end_marker: str
                 subsringer: SmartSubstringer):
        self.root_dir: str = root_dir
        self.trim: Union[str, None] = trim
        self.recursive: bool = recursive
        self.subsringer: SmartSubstringer = subsringer


class IncludeFilePlugin(Md2HtmlPlugin):

    def __init__(self):
        super().__init__()
        self.data: Dict[str, IncludeFileData] = {}
        self.all_metadata_handlers = None
        with open(MODULE_DIR.joinpath('include_file_metadata_schema.json'), 'r',
                  encoding="utf-8") as schema_file:
            self.metadata_schema = json.load(schema_file)

    def accept_data(self, data):
        self.assure_accept_data_once()
        self.validate_data_with_file(data, MODULE_DIR.joinpath('include_file_schema.json'))

        for item in data:
            for marker in item["markers"]:
                marker = marker.upper()
                if marker in self.data:
                    raise UserError(f"Marker duplication (case-insensitively): {marker}")
                substringer = SmartSubstringer(start_with=item.get("start-with", ""),
                                               end_with=item.get("end-with", ""),
                                               start_marker=item.get("start-marker", ""),
                                               end_marker=item.get("end-marker", ""),
                                               )
                self.data[marker] = IncludeFileData(root_dir=item["root-dir"],
                                                    trim=item.get("trim", "all"),
                                                    recursive=item.get("recursive", False),
                                                    subsringer=substringer,
                                                    )

    def is_blank(self) -> bool:
        return not bool(self.data)

    def accept_app_data(self, plugins: list, options: Options, metadata_handlers: PageMetadataHandlers):
        self.all_metadata_handlers = metadata_handlers

    def page_metadata_handlers(self):
        return [(self, marker, False) for marker in self.data.keys()]

    def accept_page_metadata(self, doc: Document, marker: str, metadata_str: str, metadata_section,
                             visited_markers: Union[Dict[str, None]] = None):

        marker_data = self.data[marker]

        subsringer = marker_data.subsringer
        try:
            metadata = dict_from_string_or_object(metadata_str.strip(), "file", self.metadata_schema)
        except UserError as e:
            raise UserError(f"Error in inclusion: {str(e)}")

        subsringer = subsringer.smart_copy(
            metadata.get("start-with"),
            metadata.get("end-with"),
            metadata.get("start-marker"),
            metadata.get("end-marker"),
        )

        file_path = metadata.get("file").strip()
        include_file = Path(marker_data.root_dir).joinpath(file_path)
        try:
            content = read_lines_from_cached_file(include_file)
        except FileNotFoundError as e:
            raise UserError(f"Error processing page metadata block: {type(e).__name__}: {e}")

        content = subsringer.substring(content)

        trim = metadata.get("trim", marker_data.trim)
        if trim == "all":
            content = content.strip()
        elif trim == "empty-lines":
            content = strip_empty_lines(content)

        recursive = metadata.get("recursive", marker_data.recursive)
        if recursive:
            content = apply_metadata_handlers(content, self.all_metadata_handlers, doc,
                                              visited_markers=visited_markers,
                                              recursive_marker=f"INCLUDE_FILE_PLUGIN:{include_file}")

        return content
